dhis-2/dhis-services/dhis-service-analytics/src/test/java/org/hisp/dhis/analytics/event/data/EventAnalyticsServiceTest.java

```diff
@@ -145,100 +132,202 @@ public void setUpTest()
+            EventQueryParams params = entry.getValue();
 
-    @Override
-    public void tearDownTest()
-    {
-        analyticsTableGenerator.dropTables();
+            // When
+            Grid aggregatedDataValueGrid = eventAnalyticsService.getAggregatedEventData( params );
+
+            // Then
+            AnalyticsTestUtils.assertResultGrid( aggregatedDataValueGrid, results.get( key ) );
+        }
     }
 
     @Test
-    public void testGridAggregation()
+    public void testDimensionRestrictionSuccessfully()
     {
+        // Given
+
+        // A program
+        Program aProgram = createProgram( 'B', null, null, Sets.newHashSet( ouA, ouB ), null );
+        aProgram.setUid( "aProgram123" );
+        idObjectManager.save( aProgram );
+
+        // The category options
+        CategoryOption coA = createCategoryOption( 'A' );
+        CategoryOption coB = createCategoryOption( 'B' );
+        categoryService.addCategoryOption( coA );
+        categoryService.addCategoryOption( coB );
+
+        // The categories
+        Category caA = createCategory( 'A', coA );
+        Category caB = createCategory( 'B', coB );
+        categoryService.addCategory( caA );
+        categoryService.addCategory( caB );
+
+        // The constraints
+        Set<Category> catDimensionConstraints = Sets.newHashSet( caA, caB );
+
+        // The user
+        User user = createUser( "A", "F_VIEW_EVENT_ANALYTICS" );
+        user.getUserCredentials().setCatDimensionConstraints( catDimensionConstraints );
+        userService.addUser( user );
+        enableDataSharing( user, aProgram, AccessStringHelper.DATA_READ_WRITE );
+        idObjectManager.update( user );
+        injectSecurityContext( user );
+
+        // All events in program B - 2017
+        EventQueryParams events_2017_params = new EventQueryParams.Builder()
+            .withOrganisationUnits( Lists.newArrayList( ouA ) )
+            .withStartDate( getDate( 2017, 1, 1 ) )
+            .withEndDate( getDate( 2017, 12, 31 ) )
+            .withProgram( aProgram )
+            .build();
+
+        // The params
+        Map<String, EventQueryParams> eventQueryParams = new HashMap<>();
+        eventQueryParams.put( "events_2017", events_2017_params );
+
+        // The results
+        Map<String, Double> events_2017_keyValue = new HashMap<>();
+        events_2017_keyValue.put( "ouabcdefghA", 6.0 );
+
+        Map<String, Map<String, Double>> results = new HashMap<>();
+        results.put( "events_2017", events_2017_keyValue );
+
         for ( Map.Entry<String, EventQueryParams> entry : eventQueryParams.entrySet() )
         {
             String key = entry.getKey();
             EventQueryParams params = entry.getValue();
 
+            // When
             Grid aggregatedDataValueGrid = eventAnalyticsService.getAggregatedEventData( params );
 
+            // Then
             AnalyticsTestUtils.assertResultGrid( aggregatedDataValueGrid, results.get( key ) );
```

Comment:
I agree with your reasons for wanting independent unit tests. I'm glad the you reduced the test time by only running `analyticsTableGenerator.generateTables` once.

However I believe that `analyticsTableGenerator.generateTables` is needed for the logic in both `testGridAggregation` and `testDimensionRestrictionSuccessfully`. The latter doesn't rebuild the analytics tables now, but calls `AnalyticsTestUtils.assertResultGrid` to match the results with expected data. Unfortunately there is a serious flaw in `assertResultGrid`. It checks that any results are in what is expected, but it doesn't check that anything expected is in the results. Therefore the check succeeds when passed empty results. Since you are not rebuilding the analytics tables, I suspect that it succeeds only because you pass it an analytics grid with no data.

So if we want self-contained unit tests, I think we need to build the analytics tables multiple times. For the current functionality this means 2 times, but this would increase if additional tests are added in the future.

Conversely, if we want only a single analytics tables build, I think we need some way of combining the tests that need the analytics tables into a single JUnit test--such as the way `AnalyticsServiceTest` and `EventAnalyticsServiceTest` were designed.

Because of the analytics build time (and because we might add further tests in the future that depend on the analytics build) I would still recommend a single analytics build, even though I agree with your reasons for preferring independent unit tests. But I dislike both options -- a longer test time vs. less readable test code. I would welcome other opinions. @jbee would you like to express an opinion on this? See also @maikelarabori's response to my previous comments.